From c6c326dbfdd7717ef558418d04eb1236639e255a Mon Sep 17 00:00:00 2001
From: Jason Lowe <jlowe@apache.org>
Date: Tue, 2 Feb 2016 19:30:45 +0000
Subject: [PATCH 1352/2848] MAPREDUCE-6621. Memory Leak in
 JobClient#submitJobInternal(). Contributed by
 Xuan Gong (cherry picked from commit
 43e669b22d9ac0b81ab8cc397c5d6d92bac986c6)

Change-Id: If7799c6a97752e529604eb27c5c9bb88ec257d01
---
 .../java/org/apache/hadoop/mapred/JobClient.java   |    8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-core/src/main/java/org/apache/hadoop/mapred/JobClient.java b/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-core/src/main/java/org/apache/hadoop/mapred/JobClient.java
index 752574b..c3d337e 100644
--- a/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-core/src/main/java/org/apache/hadoop/mapred/JobClient.java
+++ b/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-core/src/main/java/org/apache/hadoop/mapred/JobClient.java
@@ -579,10 +579,18 @@ public Job run() throws IOException, ClassNotFoundException,
           return job;
         }
       });
+
+      Cluster prev = cluster;
       // update our Cluster instance with the one created by Job for submission
       // (we can't pass our Cluster instance to Job, since Job wraps the config
       // instance, and the two configs would then diverge)
       cluster = job.getCluster();
+
+      // It is important to close the previous cluster instance
+      // to cleanup resources.
+      if (prev != null) {
+        prev.close();
+      }
       return new NetworkedJob(job);
     } catch (InterruptedException ie) {
       throw new IOException("interrupted", ie);
-- 
1.7.9.5

